secboot/secboot_tpm.go

```diff
@@ -273,9 +276,73 @@ func UnlockVolumeUsingSealedKeyIfEncrypted(disk disks.Disk, name string, sealedE
 	}
 }
 
+// FDERevealKeyRequest carries the operation and parameters for the
+// fde-reveal-key binary to support unsealing keys that were sealed
+// with the "fde-setup" hook.
+type FDERevealKeyRequest struct {
+	Op string `json:"op"`
+
+	SealedKey     []byte `json:"sealed-key"`
+	SealedKeyName string `json:"sealed-key-name"`
+
+	// TODO: add VolumeName,SourceDevicePath later
+}
+
+// fdeRevealKeyRuntimeMax is the maximum runtime a fde-reveal-key can execute
+// XXX: what is a reasonable default here?
+var fdeRevealKeyRuntimeMax = "2m"
+
+// fdeRevealKeyCommand returns a *exec.Cmd that is suitable to run
+// fde-reveal-key using systemd-run
+func fdeRevealKeyCommand() *exec.Cmd {
+	return exec.Command(
+		"systemd-run",
+		"--pipe", "--same-dir", "--wait", "--collect",
+		"--service-type=exec",
+		"--quiet",
+		// ensure we get some result from the hook within a
+		// reasonable timeout and output from systemd if
+		// things go wrong
+		fmt.Sprintf("--property=RuntimeMaxSec=%s", fdeRevealKeyRuntimeMax),
+		`--property=ExecStopPost=/bin/sh -c 'if [ "$EXIT_STATUS" != 0 ]; then echo "service result: $SERVICE_RESULT" 1>&2; fi'`,
+		// do not allow mounting, this ensures hooks in initrd
+		// can not mess around with ubuntu-data
+		"--property=SystemCallFilter=~@mount",
```

Comment:
to be clear, this is not a strict enforcement, there are still things that this hook could do to "escape" the inability to use the mount syscall directly such as creating another systemd service which does that mount on it's behalf with i.e. systemd-mount right?

given that I'm wondering if it's really useful to try and limit the syscalls here if we don't have a full confinement story here.

Reply:
or maybe this is just intended to minimize the number of foot shootings?

Reply:
Thanks for your feedback, I think I should expand the comment here a bit. This is really just meant to signal that the fde-reveal-key hook should not be abused to do stuff on ubuntu-data. It's a tempting place to add hacks that mess around with ubuntu-data and we just want to draw a line. We can't enforce that because and ultimately this is part of the kernel snap so if someone wants to do hacks we cannot stop this. But it should send a message: "this hook is really for unsealing, nothing else".

Reply:
Okay that's fair

Reply:
Thanks, I updated the comment